### DISCUSSION TITLE
state-machine: what do we mean with "processed"?

### DISCUSSION THREAD

**Discussion by ndr-brt** [eclipse-edc/Connector#3521] [General]
Currently the functions that are executed by the state machines on every iteration return a `boolean`.
Those boolean are counted by `ProcessorImpl`:
https://github.com/eclipse-edc/Connector/blob/9d39f5cffd8cc8b08ec8909708233943a5396e5b/core/common/state-machine/src/main/java/org/eclipse/edc/statemachine/ProcessorImpl.java#L61-L62

and the count is used to decide whether to add a delay on the iteration or not:
https://github.com/eclipse-edc/Connector/blob/9d39f5cffd8cc8b08ec8909708233943a5396e5b/core/common/state-machine/src/main/java/org/eclipse/edc/statemachine/StateMachineManager.java#L108-L114

I discovered that there is no function that's returning `false` so we're not using that feature that, thinking about it, is not really necessary anyway because an entity that gets fetched is also supposed to be processed in a certain way. 
So the logic should/could be:
- in every iteration:
  - if no entities are fetched (thus processed) apply waitStrategy
  - if one or more entities are fetched (thus processed), do not apply waitStrategy and start next iteration immediately.

This will simplify the process functions signature that often causes headache during the implementation (should I return true or false?).

Any thoughts on that?

**Comment by paullatzelsperger** [comment]
One thing to keep in mind is, that it is possible, that an entity gets fetched, because it's not leased, but no action can be performed on it. 
Cases like this - if they exist - shouldn't contribute to the `processed` count, because it would put unnecessary load on the database.

**Reply by ndr-brt** [reply]
I don't think we should have cases where the action cannot be performed (and in fact we don't have them currently), because they will end up wasting state machine cycles. 
The cases where the `return false` was used it were for not-optimal design choices, like the status checker. 
Those kind of patterns, like a poll system that should work when an entity is in a specific state, should be implemented through an external component that could be a state machine as well (like how we did with the `PolicyMonitor`).

**Reply by paullatzelsperger** [reply]
my point exactly. if such cases don't exist, all the better.

**Reply by ndr-brt** [reply]
ok, I got your comment in the wrong way, all good! :raised_hands:

**Comment by jimmarino** [comment]
ah good point
